File: src/main/java/eu/rekawek/coffeegb/gui/SwingGui.java
package eu.rekawek.coffeegb.gui;

import eu.rekawek.coffeegb.Gameboy;
import eu.rekawek.coffeegb.CartridgeOptions;
import eu.rekawek.coffeegb.debug.Console;
import eu.rekawek.coffeegb.memory.cart.Cartridge;
import eu.rekawek.coffeegb.serial.SerialEndpoint;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Properties;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class SwingGui {

    private static final Logger LOG = LoggerFactory.getLogger(SwingGui.class);

    private static final File PROPERTIES_FILE = new File(new File(System.getProperty("user.home")), ".coffeegb.properties");

    private static final int SCALE = 2;

    private final RecentRoms recentRoms;

    private final SwingDisplay display;

    private final SwingController controller;

    private final AudioSystemSoundOutput sound;

    private final CartridgeOptions options;

    private final File initialRom;

    private final Properties properties;

    private final Console console;

    private JFrame mainWindow;

    private JCheckBoxMenuItem pauseGame;

    private Cartridge cart;

    private Gameboy gameboy;

    private boolean isRunning;

    public SwingGui(CartridgeOptions options, boolean debug, File initialRom) throws IOException {
        this.options = options;
        this.initialRom = initialRom;
        properties = loadProperties();
        recentRoms = new RecentRoms(properties);
        display = new SwingDisplay(SCALE, false);
        controller = new SwingController(properties);
        sound = new AudioSystemSoundOutput();
        console = debug ? new Console() : null;
    }

    public void run() throws Exception {
        System.setProperty("apple.awt.application.name", "Coffee GB");
        System.setProperty("sun.java2d.opengl", "true");
        SwingUtilities.invokeLater(this::startGui);
    }

    public void startEmulation(File rom) {
        Cartridge newCart;
        try {
            newCart = loadRom(rom);
        } catch (IOException e) {
            JOptionPane.showMessageDialog(mainWindow, e.getMessage(), "Can't load ROM", JOptionPane.ERROR_MESSAGE);
            return;
        }
        if (isRunning) {
            stopEmulation();
        }
        cart = newCart;
        mainWindow.setTitle("Coffee GB: " + cart.getTitle());
        gameboy = new Gameboy(cart, display, controller, sound, SerialEndpoint.NULL_ENDPOINT, console);
        gameboy.registerTickListener(new TimingTicker());
        if (console != null) {
            console.setGameboy(gameboy);
        }
        new Thread(display).start();
        new Thread(sound).start();
        new Thread(gameboy).start();
        isRunning = true;
        pauseGame.setEnabled(true);
        pauseGame.setState(false);
    }

    private void stopEmulation() {
        isRunning = false;
        if (gameboy != null) {
            gameboy.stop();
            gameboy = null;
        }
        if (cart != null) {
            cart.flushBattery();
            cart = null;
        }
        sound.stopThread();
        display.stop();
        if (console != null) {
            console.setGameboy(null);
        }
    }

    private Cartridge loadRom(File rom) throws IOException {
        Cartridge.GameboyType type = Cartridge.GameboyType.AUTOMATIC;
        if (options.isForceCgb()) {
            type = Cartridge.GameboyType.FORCE_CGB;
        }
        if (options.isForceDmg()) {
            type = Cartridge.GameboyType.FORCE_DMG;
        }
        return new Cartridge(rom, options.isSupportBatterySaves(), type, options.isUsingBootstrap());
    }

    private void startGui() {
        display.setPreferredSize(new Dimension(160 * SCALE, 144 * SCALE));

        mainWindow = new JFrame("Coffee GB");

        final JFileChooser fc = new JFileChooser();
        if (properties.containsKey("rom_dir")) {
            fc.setCurrentDirectory(new File(properties.getProperty("rom_dir")));
        }

        JMenuBar menuBar = new JMenuBar();
        mainWindow.setJMenuBar(menuBar);

        JMenu fileMenu = new JMenu("File");
        menuBar.add(fileMenu);

        JMenuItem load = new JMenuItem("Load ROM");
        fileMenu.add(load);

        JMenu recentRomsMenu = new JMenu("Recent ROMs");
        fileMenu.add(recentRomsMenu);

        load.addActionListener(actionEvent -> {
            int code = fc.showOpenDialog(load);
            if (code == JFileChooser.APPROVE_OPTION) {
                File rom = fc.getSelectedFile();
                properties.setProperty("rom_dir", rom.getParent());
                launchRom(recentRomsMenu, rom);
            }
        });

        updateRecentRoms(recentRomsMenu);

        JMenu gameMenu = new JMenu("Game");
        menuBar.add(gameMenu);

        pauseGame = new JCheckBoxMenuItem("Pause", false);
        pauseGame.setEnabled(false);
        pauseGame.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_SPACE, 0));
        gameMenu.add(pauseGame);
        pauseGame.addActionListener(actionEvent -> {
            if (gameboy != null) {
                gameboy.setPaused(pauseGame.getState());
            }
        });

        JMenu audioMenu = new JMenu("Audio");
        menuBar.add(audioMenu);

        JCheckBoxMenuItem enableSound = new JCheckBoxMenuItem("Enable", true);
        audioMenu.add(enableSound);

        enableSound.addActionListener(actionEvent -> sound.setEnabled(enableSound.getState()));

        mainWindow.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
        mainWindow.addWindowListener(new java.awt.event.WindowAdapter() {
            @Override
            public void windowClosed(java.awt.event.WindowEvent windowEvent) {
                stopGui();
            }
        });
        mainWindow.setLocationRelativeTo(null);

        mainWindow.setContentPane(display);
        mainWindow.setResizable(false);
        mainWindow.setVisible(true);
        mainWindow.pack();
        mainWindow.addKeyListener(controller);
        if (console != null) {
            new Thread(console).start();
        }
        if (initialRom != null) {
            startEmulation(initialRom);
        }
    }

    private void stopGui() {
        stopEmulation();
        if (console != null) {
            console.stop();
        }
        System.exit(0);
    }

    private static Properties loadProperties() throws IOException {
        Properties props = new Properties();
        if (PROPERTIES_FILE.exists()) {
            try (FileReader reader = new FileReader(PROPERTIES_FILE)) {
                props.load(reader);
            }
        }
        return props;
    }

    private void saveProperties() {
        try (FileWriter writer = new FileWriter(PROPERTIES_FILE)) {
            properties.store(writer, "");
        } catch (IOException e) {
            LOG.error("Can't store properties", e);
        }
    }

    private void launchRom(JMenu recentRomsMenu, File rom) {
        recentRoms.addRom(rom.getAbsolutePath());
        saveProperties();
        updateRecentRoms(recentRomsMenu);
        startEmulation(rom);
    }

    private void updateRecentRoms(JMenu recentRomsMenu) {
        recentRomsMenu.removeAll();
        for (String romPath : recentRoms.getRoms()) {
            File rom = new File(romPath);
            JMenuItem item = new JMenuItem(rom.getName());
            item.addActionListener(actionEvent -> {
                launchRom(recentRomsMenu, rom);
            });
            recentRomsMenu.add(item);
        }
    }
}

File: src/main/java/eu/rekawek/coffeegb/gui/SwingGui.kt
package eu.rekawek.coffeegb.gui

import eu.rekawek.coffeegb.CartridgeOptions
import eu.rekawek.coffeegb.Gameboy
import eu.rekawek.coffeegb.debug.Console
import eu.rekawek.coffeegb.memory.cart.Cartridge
import eu.rekawek.coffeegb.memory.cart.Cartridge.GameboyType
import eu.rekawek.coffeegb.serial.SerialEndpoint
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.event.ActionEvent
import java.awt.event.KeyEvent
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.io.File
import java.io.FileReader
import java.io.FileWriter
import java.io.IOException
import java.util.*
import javax.swing.*

class SwingGui(private val options: CartridgeOptions, debug: Boolean, private var currentRom: File?) {
    private val recentRoms: RecentRoms

    private val display: SwingDisplay

    private val controller: SwingController

    private val sound: AudioSystemSoundOutput

    private val properties: Properties

    private val console: Console?

    private var mainWindow: JFrame? = null

    private var pauseGame: JCheckBoxMenuItem? = null

    private var resetGame: JMenuItem? = null

    private var cart: Cartridge? = null

    private var gameboy: Gameboy? = null

    private var isRunning = false

    private var type: GameboyType

    init {
        properties = loadProperties()
        recentRoms = RecentRoms(properties)

        val displayScale = properties.getProperty("display.scale", "2").toInt()
        val displayGrayscale = properties.getProperty("display.grayscale", "false").toBoolean()
        val soundEnabled = properties.getProperty("sound.enabled", "true").toBoolean()
        type = GameboyType.valueOf(properties.getProperty("gameBoy.type", GameboyType.AUTOMATIC.name))

        display = SwingDisplay(displayScale, displayGrayscale)
        controller = SwingController(properties)
        sound = AudioSystemSoundOutput()
        sound.isEnabled = soundEnabled
        console = if (debug) Console() else null
    }

    @Throws(Exception::class)
    fun run() {
        System.setProperty("apple.awt.application.name", "Coffee GB")
        System.setProperty("sun.java2d.opengl", "true")
        SwingUtilities.invokeLater { this.startGui() }
    }

    fun startEmulation(rom: File?) {
        val newCart: Cartridge
        try {
            newCart = loadRom(rom)
        } catch (e: IOException) {
            JOptionPane.showMessageDialog(mainWindow, e.message, "Can't load ROM", JOptionPane.ERROR_MESSAGE)
            return
        }
        stopEmulation()
        cart = newCart
        mainWindow!!.title = "Coffee GB: " + cart!!.title
        gameboy = Gameboy(cart, display, controller, sound, SerialEndpoint.NULL_ENDPOINT, console)
        gameboy!!.registerTickListener(TimingTicker())
        console?.setGameboy(gameboy)
        Thread(display).start()
        Thread(sound).start()
        Thread(gameboy).start()
        isRunning = true
        pauseGame!!.isEnabled = true
        pauseGame!!.state = false
        resetGame!!.isEnabled = true
        currentRom = rom
    }

    private fun stopEmulation() {
        if (!isRunning) {
            return
        }
        isRunning = false
        if (gameboy != null) {
            gameboy!!.stop()
            gameboy = null
        }
        if (cart != null) {
            cart!!.flushBattery()
            cart = null
        }
        sound.stopThread()
        display.stop()
        console?.setGameboy(null)
        pauseGame!!.isEnabled = false
        pauseGame!!.state = false
        resetGame!!.isEnabled = false
    }

    @Throws(IOException::class)
    private fun loadRom(rom: File?): Cartridge {
        return Cartridge(rom, options.isSupportBatterySaves, type, options.isUsingBootstrap)
    }

    private fun startGui() {
        display.scale = display.scale
        mainWindow = JFrame("Coffee GB")

        val fc = JFileChooser()
        if (properties.containsKey("rom_dir")) {
            fc.currentDirectory = File(properties.getProperty("rom_dir"))
        }

        val menuBar = JMenuBar()
        mainWindow!!.jMenuBar = menuBar

        val fileMenu = JMenu("File")
        menuBar.add(fileMenu)

        val load = JMenuItem("Load ROM")
        fileMenu.add(load)

        val recentRomsMenu = JMenu("Recent ROMs")
        fileMenu.add(recentRomsMenu)

        load.addActionListener { actionEvent: ActionEvent? ->
            val code = fc.showOpenDialog(load)
            if (code == JFileChooser.APPROVE_OPTION) {
                val rom = fc.selectedFile
                properties.setProperty("rom_dir", rom.parent)
                launchRom(recentRomsMenu, rom)
            }
        }

        updateRecentRoms(recentRomsMenu)

        val gameMenu = JMenu("Game")
        menuBar.add(gameMenu)

        pauseGame = JCheckBoxMenuItem("Pause", false)
        pauseGame!!.isEnabled = false
        pauseGame!!.accelerator = KeyStroke.getKeyStroke(KeyEvent.VK_SPACE, 0)
        gameMenu.add(pauseGame)
        pauseGame!!.addActionListener { actionEvent: ActionEvent? ->
            if (gameboy != null) {
                gameboy!!.setPaused(pauseGame!!.state)
            }
        }

        val typeMenu = JMenu("GameBoy type")
        gameMenu.add(typeMenu)

        for (type in GameboyType.entries.toTypedArray()) {
            val item = JCheckBoxMenuItem(type.label, type == this.type)
            item.addActionListener { actionEvent: ActionEvent? ->
                this.type = type
                if (isRunning) {
                    startEmulation(currentRom)
                }
                uncheckAllBut(typeMenu, item)
                properties.setProperty("gameBoy.type", type.name)
                saveProperties()
            }
            typeMenu.add(item)
        }

        resetGame = JMenuItem("Reset")
        gameMenu.add(resetGame)
        resetGame!!.addActionListener { actionEvent: ActionEvent? ->
            if (currentRom != null) {
                startEmulation(currentRom)
            }
        }
        resetGame!!.isEnabled = false

        val screenMenu = JMenu("Screen")
        menuBar.add(screenMenu)

        val scale = JMenu("Scale")
        screenMenu.add(scale)

        for (s in mutableListOf<Int>(1, 2, 4)) {
            val item = JCheckBoxMenuItem(s.toString() + "x", s == display.scale)
            item.addActionListener { actionEvent: ActionEvent? ->
                display.scale = s
                mainWindow!!.pack()
                uncheckAllBut(scale, item)
                properties.setProperty("display.scale", s.toString())
                saveProperties()
            }
            scale.add(item)
        }

        val grayscale = JCheckBoxMenuItem("DMG grayscale", display.isGrayscale)
        screenMenu.add(grayscale)
        grayscale.addActionListener { actionEvent: ActionEvent? ->
            display.isGrayscale = grayscale.state
            properties.setProperty("display.grayscale", grayscale.state.toString())
            saveProperties()
        }

        val audioMenu = JMenu("Audio")
        menuBar.add(audioMenu)

        val enableSound = JCheckBoxMenuItem("Enable", sound.isEnabled)
        audioMenu.add(enableSound)

        enableSound.addActionListener { actionEvent: ActionEvent? ->
            sound.isEnabled = enableSound.state
            properties.setProperty("sound.enabled", enableSound.state.toString())
            saveProperties()
        }

        mainWindow!!.defaultCloseOperation = JFrame.DISPOSE_ON_CLOSE
        mainWindow!!.addWindowListener(object : WindowAdapter() {
            override fun windowClosed(windowEvent: WindowEvent) {
                stopGui()
            }
        })
        mainWindow!!.setLocationRelativeTo(null)

        mainWindow!!.contentPane = display
        mainWindow!!.isResizable = false
        mainWindow!!.isVisible = true
        mainWindow!!.pack()
        mainWindow!!.addKeyListener(controller)
        if (console != null) {
            Thread(console).start()
        }
        if (currentRom != null) {
            startEmulation(currentRom)
        }
    }

    private fun stopGui() {
        stopEmulation()
        console?.stop()
        System.exit(0)
    }

    private fun saveProperties() {
        try {
            FileWriter(PROPERTIES_FILE).use { writer ->
                properties.store(writer, "")
            }
        } catch (e: IOException) {
            LOG.error("Can't store properties", e)
        }
    }

    private fun launchRom(recentRomsMenu: JMenu, rom: File) {
        recentRoms.addRom(rom.absolutePath)
        saveProperties()
        updateRecentRoms(recentRomsMenu)
        startEmulation(rom)
    }

    private fun updateRecentRoms(recentRomsMenu: JMenu) {
        recentRomsMenu.removeAll()
        for (romPath in recentRoms.roms) {
            val rom = File(romPath)
            val item = JMenuItem(rom.name)
            item.addActionListener { actionEvent: ActionEvent? ->
                launchRom(recentRomsMenu, rom)
            }
            recentRomsMenu.add(item)
        }
    }

    companion object {
        private val LOG: Logger = LoggerFactory.getLogger(SwingGui::class.java)

        private val PROPERTIES_FILE = File(File(System.getProperty("user.home")), ".coffeegb.properties")

        @Throws(IOException::class)
        private fun loadProperties(): Properties {
            val props = Properties()
            if (PROPERTIES_FILE.exists()) {
                FileReader(PROPERTIES_FILE).use { reader ->
                    props.load(reader)
                }
            }
            return props
        }

        private fun uncheckAllBut(parent: JMenu, selectedItem: JCheckBoxMenuItem) {
            for (i in 0 until parent.itemCount) {
                val item = parent.getItem(i)
                if (item === selectedItem) {
                    continue
                }
                (item as JCheckBoxMenuItem).state = false
            }
        }
    }
}
